### PR TITLE
refactor: reduce complexity and prevent new Radon violations

### PR DESCRIPTION
**Pull request by Silas-Asamoah**
## Summary

All scoped Stormlog functions now meet the CC 10 budget, with no legacy exceptions left. CI and a zero-exception regression test prevent that debt from returning.

Profiled first with Radon 6.0.1, then reduced all 23 D/E hotspots and all 109 C-grade functions. This includes the 95 exceptions previously deferred in this PR, completed in 11 additional atomic commits. Public behavior is intended to remain unchanged.

## Measurements

| Callable metric | Before this PR | Current head |
| --- | ---: | ---: |
| C grade (CC 11–20) | 109 | 0 |
| D grade (CC 21–30) | 19 | 0 |
| E grade (CC 31–40) | 4 | 0 |

The final gate scans **1,710 callables: 1,263 A and 447 B**, with zero violations or grace entries. Counts include nested functions and methods, not class aggregates.

## Guard and compatibility decisions

- The budget covers `stormlog/` and the checker itself; tests and examples remain excluded.
- The exception map must stay empty; the limit remains 10.
- Baseline pruning remains backward-compatible but cannot add new debt.
- The checker retains exit 0 for success, 1 for violations, and 2 for invalid input.
- Cohesive private helpers preserve validation order, output ordering, and cleanup ownership; no public signatures were intentionally changed.
- AST fingerprint compatibility remains tested across Python 3.10, 3.12, 3.13, and 3.14.

The short-soak RSS integration test now uses deterministic harness readings while retaining the real CPU tracker and retention checks. Separate tests verify daily-growth normalization, transient peaks, and budget acceptance/rejection. Production budgets were not relaxed.

## Validation

Evidence below is for head `c23750f`; paired runtime checks use prior `release/dev` build `4986a35`.

- Full local suite: **1,119 passed, 20 skipped**.
- Complexity gate tests: **17 passed**.
- All pre-commit hooks pass, including Black, isort, Flake8, and mypy.
- [Current-head GitHub CI](https://github.com/Silas-Asamoah/stormlog/actions/runs/34219526211): **20 Actions checks passed, 2 conditional jobs skipped**; CodeRabbit status is successful.
- Complexity CI passed on Python 3.10, 3.12, 3.13, and 3.14.
- Both builds passed the CPU/real-MPS capability matrix: **5 passed, 2 skipped, 0 failed**.
- Deterministic sessions match: CPU **12 events**, MPS **14 events** per session.
- Duration and SIGINT CLI tracking exit **0**, finish as `completed`, and close collectors, flush threads, and sinks.
- All **13 shared-input CLI replay cases** match, including exit **1** for malformed or missing input.
- Event and rollup artifacts validate against both builds' schemas; CSV headers match.
- Diagnose CLI and real PTY TUI startup/quit exit **0** on both builds.
- No new collector failures or unexpected tracebacks were observed.

Separate live CPU workloads produced one prior peak event versus two candidate peak events; sample/allocation/deallocation counts matched. Fixed-input lifecycle and replay checks matched. This is not a claim that independently sampled live RSS values must be identical.

## Limitations

The latest batch received a manual compatibility-focused diff scan, not a completed independent automated review.

Code review: skipped (ce-code-review unavailable) — all local persona dispatches failed with unsupported model; the peer adapter rejected --safe-mode. Explicit manual diff scan and compatibility probes replace the unavailable dedicated review, not an independent-review claim.

Local validation did not exercise real CUDA/ROCm, native CUDA OOM capture, multi-GPU behavior, live TensorFlow/JAX runtimes, or a production-duration soak. The OOM smoke used a simulated exception. CI's PyTorch, TensorFlow, and JAX test jobs passed; that is not a substitute for untested GPU hardware. This is pre-merge validation, not evidence of a deployment.

An earlier CPU-only parser microbenchmark measured approximately 1.8 microseconds per record of additional overhead (+6.7%). It was not rerun for this latest batch and is not an end-to-end throughput measurement.

## Post-Deploy Monitoring & Validation

For the first `release/dev` validation and hardware smoke run, the PR author/maintainer should compare CLI exit codes, session status, event counts, and `collector_failure`/traceback logs with the prior build. Healthy signals are successful collector cleanup, unchanged artifact schemas, and no new failures. Reproduce any output or lifecycle regression with the same input and revert the relevant atomic refactor before release if confirmed.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

### CODE REVIEW

File: scripts/check_complexity.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+def _source_hash(node: ast.AST) -> str:
+    normalized = copy.deepcopy(node)
+    for child in ast.walk(normalized):
+        # Python 3.12 adds this empty field even to pre-3.12 syntax.
+        if getattr(child, "type_params", None) == []:
+            delattr(child, "type_params")
+    return hashlib.sha256(
+        ast.dump(normalized, include_attributes=False).encode("utf-8")
+    ).hexdigest()
```
[the remark this answers is not in the document]
Comment (Silas-Asamoah) [reply scripts/check_complexity.py:65]: > serialize with consistent empty-field behavior across versions

Fixed in a56cfcc. The gate now explicitly uses `show_empty=True` on Python 3.13+, preserving the original fingerprints without changing the baseline. Added a fixed legacy-baseline regression test and expanded the complexity CI matrix to Python 3.10, 3.12, 3.13, and 3.14.

Verified identical full-baseline results on all four versions (including Python 3.14.7), with 25 gate/workflow tests passing per version. The main suite passed 1,021 tests, and all commit hooks passed.
